### Classification approach for exercise

In [1]:
import os,re
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
from sklearn import metrics
from sklearn.metrics import accuracy_score
from sklearn.metrics import mean_absolute_error,mean_squared_error
from sklearn.metrics import classification_report,confusion_matrix,log_loss
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier

pd.set_option('display.max_columns',200)
pd.set_option("display.max_rows",500)

In [2]:
dfTrain = pd.read_csv("exercise_40_train.csv")
dfTest = pd.read_csv("exercise_40_test.csv")
dfTest.head(15)


x1         x2         x3        x4        x5        x6        x7  \
0   4.747627  20.509439  Wednesday  2.299105 -1.815777 -0.752166   0.0098%   
1   1.148654  19.301465        Fri  1.862200 -0.773707 -1.461276   0.0076%   
2   4.986860  18.769675   Saturday  1.040845 -1.548690  2.632948   -5e-04%   
3   3.709183  18.374375    Tuesday -0.169882 -2.396549 -0.784673   -0.016%   
4   3.801616  20.205541     Monday  2.092652 -0.732784 -0.703101   0.0186%   
5   1.449185  17.237848        Sun  0.998924  0.882350 -1.700730   0.0076%   
6   0.979035  18.056038     Monday  0.105417 -1.782850 -1.390759  -0.0045%   
7   0.910023  17.057286        Mon -0.903745  0.118699 -2.110974  -0.0079%   
8   3.616287  17.740881       Thur  2.193121  4.222402  0.206030  -0.0201%   
9   2.539012  21.078729   Saturday -0.868620 -0.193794  1.244369  -0.0222%   
10  2.202450  20.431475        Mon  0.639342 -2.867947 -1.407299   -3e-04%   
11  6.334361  19.658003     Friday  0.555120 -0.096129  0.861986  -0.0066%   
12  4.047808  23.034961        Sat -0.505224 -0.864426  0.115058  -0.0158%   
13  3.078519  21.370217  Wednesday -0.918152 -0.777777  0.695998  -0.0096%   
14  4.768995  21.048787  Wednesday -0.183138 -0.959152  0.501082  -0.0195%   

          x8        x9       x10         x11        x12       x13       x14  \
0  -3.240309  0.587948 -0.260721  101.113628  -0.812035  3.251085 -0.004432   
1   0.443209  0.522113 -1.090886  104.791999   8.805876  1.651993       NaN   
2  -1.167885  5.739275  0.222975  102.109546   7.831517  3.055358  2.036434   
3  -2.662226  1.548050  0.210141   82.653354   0.436885  1.578106       NaN   
4   0.056422  2.878167 -0.457618   75.036421   8.034303  1.631426  0.643738   
5  -0.371594  1.269817 -1.433127   78.434085   3.124643  4.438999       NaN   
6  -1.886515  2.905928  0.222237   87.040247   6.739816  2.839826  0.227705   
7  -0.613028  2.109169  1.140936   93.985642   1.006534  2.645002  0.938672   
8  -0.397257  3.058279 -0.859673  120.832847   8.501188  2.386375  0.841161   
9  -0.334661  1.478947  0.637950   88.678380  -1.291054  2.964086  0.268776   
10  1.538257  3.544400 -0.965467  101.354472   8.760391  5.464629  0.733497   
11  0.594773  1.582971  1.478990   97.004902  -2.995886  3.045324  0.115462   
12 -2.283415  1.985695 -1.769429  105.941368  10.299181  2.836541  0.835711   
13  0.632970  2.430308  1.253636  105.366919   3.839965  6.779266  0.994631   
14  2.752925  5.526124 -1.602605  122.870329   7.220539  2.991452       NaN   

         x15         x16        x17        x18                 x19       x20  \
0  -0.917763  102.193597   7.097387  75.256714   $120.216189955777  8.269754   
1  -1.396065   99.356609   7.117930  79.061540  $-267.562586413086  5.668347   
2   1.057296         NaN  10.943217  87.567827  $-311.292903116571  3.219583   
3  -1.287913  102.410965   6.588790  71.825782   $2229.14940030076  7.459929   
4   0.349166  101.513490   5.777599  74.602441  $-469.049529991235  8.245194   
5  -0.761186  102.214681  10.767388  77.358212   $1052.65320492622  4.483987   
6  -3.909978  100.183345   6.302176  81.401764  $-1248.56374994913  7.149937   
7  -1.301097  100.798786   5.676887  86.843134   $2192.44913918191  7.323247   
8   2.918181   99.884976   3.951736  71.032462   $-744.50837755944  8.676654   
9  -1.237706         NaN   7.867055  89.714961   $621.500064669709  5.846521   
10  0.618157         NaN  10.094763  90.123582   $1641.48095109299  0.695014   
11 -0.080587         NaN   7.524240  80.905777  $-2803.98880622137  9.902445   
12  1.468347   99.542627   4.883594  73.840790  $-960.320565434093 -0.011208   
13  0.899184         NaN  10.035408  81.874254  $-360.765613820984  5.393525   
14 -0.610510  101.378570   9.087118  78.425467  $-743.622584295803  2.311300   

          x21         x22       x23     x24        x25       x26          x27  \
0    4.794438   56.134458 -0.083641     NaN   5.740955 -3.152540   924.013304   
1   12.377178         NaN  0.321981  female   2.05812

### Combine data frames

#### Goal is to encode categorical variables for train & test

In [3]:
dfCombine = pd.concat([dfTrain,dfTest])
dfCombine.iloc[40000:40001,]

y        x1         x2         x3        x4        x5        x6       x7  \
0 NaN  4.747627  20.509439  Wednesday  2.299105 -1.815777 -0.752166  0.0098%   

         x8        x9       x10         x11       x12       x13       x14  \
0 -3.240309  0.587948 -0.260721  101.113628 -0.812035  3.251085 -0.004432   

        x15         x16       x17        x18                x19       x20  \
0 -0.917763  102.193597  7.097387  75.256714  $120.216189955777  8.269754   

        x21        x22       x23  x24       x25      x26         x27  \
0  4.794438  56.134458 -0.083641  NaN  5.740955 -3.15254  924.013304   

         x28       x29  x30 x31       x32      x33        x34     x35  \
0  17.697125  1.402273  NaN  no  1.461654  Florida  57.930285  4.7273   

         x36        x37       x38         x39        x40           x41  \
0  66.416594  28.450466  33.49931  5-10 miles  16.776061  99971.844703   

        x42       x43  x44  x45       x46        x47        x48  x49  \
0  0.492812 -0.963616  NaN  NaN  1.266416  53.020326  61.840284  NaN   

           x50       x51  x52        x53  x54       x55       x56  x57  \
0  1825.413159  2.517924  NaN  70.880778  NaN  1.923125  0.239009  NaN   

         x58  x59  x60       x61       x62       x63        x64          x65  \
0  300.62999    0  May  5.461123  5.149473  1.179229  59.346822  progressive   

        x66        x67       x68       x69       x70       x71        x72  \
0  2.219502  17.667792  5.557066 -2.030901  0.636111 -0.000234  62.363381   

        x73       x74       x75       x76       x77        x78  x79       x80  \
0  4.613878  2.415655  3.632041 -0.382482  mercedes  55.938387  1.0  8.325299   

        x81       x82      x83      x84  x85      x86       x87       x88  \
0 -1.274085 -5.663245  2.40266 -0.06104  NaN  64.0025  0.548974  3.420875   

         x89       x90  x91        x92 x93       x94       x95        x96  \
0  11.553023  5.904644  NaN  12.542333  no  3.107683  0.533904  12.438759   

        x97  x98  x99      x100  
0  7.298306    0  NaN  93.56712

In [4]:
dfCombine.reset_index(drop=True,inplace=True)
dfCombine.tail(1)

y        x1         x2      x3        x4        x5        x6       x7  \
49999 NaN  2.972268  20.485504  Sunday  1.747878  0.197357  1.798676  0.0086%   

             x8        x9      x10        x11       x12       x13       x14  \
49999 -0.318751  2.858806  0.08416  99.642148  8.497151  0.878065 -0.338883   

            x15         x16       x17       x18                 x19       x20  \
49999  0.350812  101.065927  7.900097  76.79105  $-508.049552925335  0.443446   

            x21        x22       x23     x24       x25  x26         x27  \
49999 -0.255657  82.889762 -2.112773  female  2.764069  NaN  1028.57299   

            x28       x29       x30  x31       x32  x33        x34       x35  \
49999  18.45714  2.473741 -0.578496  yes  0.634841  NaN  50.916587  1.271283   

             x36        x37        x38         x39        x40           x41  \
49999  63.837508  29.870539  39.909804  5-10 miles  19.938424  101720.09239   

            x42       x43  x44       x45       x46        x47        x48  \
49999 -0.875818  0.765511  NaN  0.996541 -3.489898  49.483851  65.862504   

            x49         x50       x51       x52        x53      x54       x55  \
49999  7.303529  1150.95107  1.808276  9.110502  72.228876 -0.61096  2.634673   

           x56  x57        x58  x59        x60       x61       x62       x63  \
49999 -1.92103  NaN  300.62999    0  September  6.847741  2.181149  1.297841   

             x64          x65       x66        x67        x68      x69  \
49999  50.885163  progressive  0.153248  14.402071  11.865657 -1.29376   

            x70       x71        x72       x73  x74       x75       x76  x77  \
49999  0.434589 -0.000234  52.004365  4.865971  NaN  1.524051  0.731896  NaN   

             x78  x79       x80       x81       x82       x83       x84  \
49999  45.800511  1.0  7.780886  0.953972 -6.183223  1.565181  3.140183   

             x85  x86       x87       x88        x89       x90       x91  \
49999  95.665994  NaN -0.127836 -1.550846  11.222902  4.651827  1.397517   

            x92  x93       x94       x95  x96        x97  x98  x99       x100  
49999  8.463199  yes  3.942066  1.245289  NaN  11.718582    0  yes  98.307681

### Clean data

#### Removal of null values

In [5]:
dfCombine.isnull().sum()

y       10000
x1          0
x2          0
x3          0
x4          0
x5       3030
x6          0
x7          0
x8          0
x9          0
x10         0
x11      6439
x12         0
x13         0
x14     12292
x15         0
x16     13965
x17         0
x18         0
x19         0
x20         0
x21         0
x22      3000
x23         0
x24      4825
x25         0
x26      3050
x27         0
x28         0
x29         0
x30     40421
x31         0
x32         0
x33      8941
x34         0
x35         0
x36         0
x37         0
x38      2967
x39         0
x40         0
x41     11907
x42     12147
x43         0
x44     42813
x45     10072
x46         0
x47         0
x48         0
x49     16077
x50         0
x51         0
x52     20262
x53         0
x54     15934
x55     22120
x56         0
x57     40541
x58         0
x59         0
x60         0
x61      9059
x62         0
x63      3007
x64      6363
x65         0
x66         0
x67      3045
x68      2984
x69         0
x70         0
x71   

### Drop columns with null values & other un-needed columns

In [6]:
# dfCombine.drop(["x5","x11","x14","x16","x22","x24","x26","x30","x33","x38","x41","x42","x44","x45","x49","x52","x54",\
# "x55","x57","x61","x63","x64","x67","x68","x74","x75","x76","x77","x78","x79","x80","x85","x89","x95","x99",\
# "x83","x86","x88","x91","x92","x94","x96"],axis=1,inplace=True)
# unneeded: x39
dfCombine.drop(["x11","x14","x16","x24","x30","x33","x38","x39","x41","x42","x44","x45","x49","x52","x54",\
"x55","x57","x61","x64","x74","x75","x76","x77","x78","x79","x80","x85","x89","x95","x99",\
"x83","x86","x88","x91","x96"],axis=1,inplace=True)

In [7]:
dfCombine.isnull().sum()

y       10000
x1          0
x2          0
x3          0
x4          0
x5       3030
x6          0
x7          0
x8          0
x9          0
x10         0
x12         0
x13         0
x15         0
x17         0
x18         0
x19         0
x20         0
x21         0
x22      3000
x23         0
x25         0
x26      3050
x27         0
x28         0
x29         0
x31         0
x32         0
x34         0
x35         0
x36         0
x37         0
x40         0
x43         0
x46         0
x47         0
x48         0
x50         0
x51         0
x53         0
x56         0
x58         0
x59         0
x60         0
x62         0
x63      3007
x65         0
x66         0
x67      3045
x68      2984
x69         0
x70         0
x71         0
x72         0
x73         0
x81         0
x82         0
x84         0
x87         0
x90         0
x92      3061
x93         0
x94      2955
x97         0
x98         0
x100        0
dtype: int64

In [8]:
dfCombine[["x5","x22","x26","x63","x67","x68","x92","x94"]].sample(30)

x5         x22       x26       x63        x67        x68  \
18980       NaN   61.313371 -0.622774       NaN  14.402071  22.649392   
6168  -0.475327   30.024490 -0.330055 -0.077186  14.402071  19.622046   
30120       NaN   33.542520 -0.842309 -0.008392        NaN  18.916159   
19958       NaN  109.770123 -0.337769       NaN  14.402071  16.989091   
17443 -3.537619   33.188748 -4.425757  0.857456  14.402071  28.031547   
46232 -1.572788   67.600785  0.895919 -1.543204  14.402071  13.304211   
43785  0.639496   79.686199 -0.470341  0.038175  14.402071  17.118425   
32524 -1.734732   67.366693 -0.370425 -0.445271  14.402071  17.610884   
20603 -0.449163   70.633959 -1.815152  1.833388  14.402071  10.791129   
18006  0.711825   49.685039  0.750973 -2.676265  11.499010  19.450227   
41433  0.759565  105.096879       NaN -1.149290  14.402071  25.587973   
48427  2.375268   35.019146 -1.272225 -1.832183  14.402071  18.299599   
29736  1.687615   97.345531  2.549405 -0.443728  14.402071  15.276070   
46602 -0.622215   90.541953 -1.670085  1.175776  14.402071   8.707415   
3558   1.477735   85.246351  1.429292  2.443562  14.402071  27.752262   
36441  1.692764   71.942159 -1.364054 -1.228587  14.402071  19.709073   
4199   0.316535   72.607701 -1.296998  1.541213  14.402071   9.709520   
41599  0.673644   53.162552 -0.573930 -3.608229  14.402071  25.769836   
34060 -0.687181   53.269555  4.695551 -1.122386        NaN  15.199126   
33062 -1.278849   43.353385 -2.696339  0.473083  14.402071        NaN   
1211        NaN   43.895991 -0.526535  3.038028  14.402071  18.852015   
27769 -0.810668   63.748765  0.555997  0.175257  14.402071  14.408472   
11407       NaN   26.512238       NaN  0.739969  14.402071  11.636815   
40634  0.036374   75.961592  1.040435  1.274288  14.402071  16.980003   
34821 -1.353692   25.585180 -0.700722 -0.511281  14.402071   9.581647   
33324 -0.712800   72.856409 -0.163009 -0.054612  14.402071  21.099938   
42092  0.065884   78.404411       NaN  1.164315  14.402071  11.438020   
38126  1.075582   88.710075  0.993948 -0.311949  14.402071   1.440283   
48215       NaN   55.541748  1.990544       NaN  14.402071   7.352103   
10715  1.117296         NaN -2.218904 -0.228768  14.402071  15.474752   

             x92       x94  
18980   6.333430  3.495313  
6168    9.315357  4.642095  
30120   5.261618  2.977995  
19958   7.200700  2.954908  
17443   5.557540  1.709614  
46232   9.549205  5.765680  
43785  -0.944080  4.168400  
32524   6.328149       NaN  
20603   0.171589  2.603015  
18006   3.728183  0.591311  
41433   5.378939  2.412637  
48427   7.307758  5.943773  
29736   1.914714  5.336639  
46602  -0.763426  3.229851  
3558         NaN  3.370540  
36441   3.866899  4.383357  
4199    9.239133  6.887503  
41599   6.326426  3.571292  
34060   4.614709  9.484059  
33062   7.969250  2.283589  
1211    3.854107  3.905143  
27769   3.720717       NaN  
11407   0.099029  4.447553  
40634   5.098519  3.204723  
34821   4.337499  4.254568  
33324  10.230932  4.510916  
42092   5.360118       NaN  
38126   5.122642  4.245587  
48215   0.964828  5.392539  
10715   5.702984  1.136304

### Interpolate missing values

In [9]:
dfCombine.interpolate(method="linear",inplace=True)
dfCombine.sample(25)

y        x1         x2         x3        x4        x5        x6  \
31781  0.0  6.152333  19.336132        Sun  0.379153  1.692065 -0.689577   
12473  0.0  0.907120  24.017640   Saturday  0.779066 -1.929447  2.424314   
19900  0.0 -0.123385  19.504144     Monday -2.695419  0.796772  0.652831   
19201  0.0  0.117723  17.239438    Tuesday -1.406980 -0.107757  0.111944   
39179  0.0  4.704678  18.148591   Thursday -0.888025  2.504513 -0.051067   
40631  0.0  1.487025  19.057590    Tuesday -1.378951  0.933801  0.359624   
9795   0.0  4.988083  16.170532        Fri -0.780454  2.075966  0.130068   
18497  0.0  0.118929  22.049082     Monday  0.143329 -0.785774 -1.014443   
40066  0.0  4.569200  21.635767   Saturday -1.256005  2.308564  0.465082   
12724  0.0  3.152499  20.725063    Tuesday -0.883072 -0.002951 -0.720985   
6080   0.0  4.343438  19.835074    Tuesday  1.947352  1.768166 -1.438391   
11528  0.0  4.760418  20.817334    Tuesday -1.524575  0.505883 -0.207342   
44187  0.0  6.531372  17.702092        Fri  0.554175  0.943919 -1.367207   
46034  0.0  7.913126  20.001633        Sat  1.569034 -1.825908  2.535338   
35692  0.0  3.036519  18.962628       Thur  1.115199 -0.237297 -1.221795   
24298  0.0  1.314586  20.100809     Friday -0.156990  0.638327  1.543130   
29294  1.0  3.633273  20.790608   Thursday  1.226313  0.286500  0.041654   
22316  0.0  0.929402  22.916828   Saturday  0.582260  1.050422  1.028953   
39168  0.0  3.225780  18.179749        Sun  0.463751 -0.624606  0.073113   
18893  0.0  1.160035  18.012988     Friday -2.540453  1.398056  1.726679   
41615  0.0  1.679740  20.258668        Sun -1.519490  1.316917 -1.563379   
25512  0.0  4.810633  22.992353     Sunday  0.552323 -0.116765 -3.428801   
22934  1.0  6.985950  17.254693  Wednesday -0.237756  1.850016 -2.434942   
5798   0.0  2.795521  17.607398     Monday -0.217170  0.089238 -0.419915   
45985  0.0  1.750529  21.283222        Sat -1.182646 -0.860831  4.083015   

             x7        x8        x9       x10       x12       x13       x15  \
31781   0.0021% -0.450959  4.253360  0.654724  7.113846  7.222348 -0.755193   
12473   0.0043% -2.353797  1.100592  0.555853  2.011547  4.981047 -0.031826   
19900   0.0085% -1.775996  2.614620  2.106164  6.453064  2.260257  0.930211   
19201  -0.0029% -0.914733  4.041597  0.738707  0.202707  5.625008 -1.123931   
39179   0.0055%  0.514483  3.580039  0.402596  2.508419  3.330504  0.308413   
40631   -0.005%  2.372394  3.448572  0.056224 -0.346889  4.903374 -0.114476   
9795    0.0193% -1.960786  1.643604 -2.169732  4.537948  6.186191 -0.189353   
18497    3e-04% -0.164258  4.352330  0.346326  8.042841  0.634736  1.168302   
40066  -0.0161%  0.953893 -0.877348 -0.322708 -0.793343  3.774738  0.015952   
12724   0.0085% -0.793783  2.274695  0.891738  0.365335  3.316052  0.592877   
6080    0.0113% -1.192567 -0.084141 -0.664271  6.446373  5.909114 -0.312639   
11528   0.0044% -1.194021  4.850478  0.672393  7.920679  8.920295  2.576097   
44187   0.0224% -0.409930  1.417099  2.849494  4.539161  2.495307 -1.053105   
46034  -0.0021%  1.063194  2.610165  1.547185 -0.813560 -2.102274 -1.271581   
35692   0.0053%  1.131110  3.620176 -0.364986  6.683483  6.703980  0.541555   
24298  -0.0154%  0.072387  4.836247 -0.288502  6.590181  3.918696  2.125918   
29294  -0.0073%  0.178227  2.332161 -0.555732  4.183504  4.029829  1.357737   
22316   0.0099%  4.861546  2.353296 -0.365942  0.719728  7.774400 -0.162473   
39168  -0.0111% -0.292420  1.996200  0.253169  4.407813  1.557951 -0.534634   
18893  -0.0199%  0.211054  4.020728  1.644898 -0.834568  4.954776  0.087551   
41615   -0.019% -0.264754  4.088041  1.533397  4.281991  5.901156  0.274013   
25512  -0.0074% -0.389734  1.312473  2.183372  3.497714  6.251170 -1.244583   
22934  -0.0102% -1.682558  2.639129 -0.906586  2.572307  4.515257  0.959550   
5798    0.0088% -1.604304  1.989127 -0.328812  1.894189  7.580578  1.478784   
45985   0.0048%  1.053475  6.708441 -1.499393  7.682112  

In [10]:
dfCombine.isnull().sum()

y       0
x1      0
x2      0
x3      0
x4      0
x5      0
x6      0
x7      0
x8      0
x9      0
x10     0
x12     0
x13     0
x15     0
x17     0
x18     0
x19     0
x20     0
x21     0
x22     0
x23     0
x25     0
x26     0
x27     0
x28     0
x29     0
x31     0
x32     0
x34     0
x35     0
x36     0
x37     0
x40     0
x43     0
x46     0
x47     0
x48     0
x50     0
x51     0
x53     0
x56     0
x58     0
x59     0
x60     0
x62     0
x63     0
x65     0
x66     0
x67     0
x68     0
x69     0
x70     0
x71     0
x72     0
x73     0
x81     0
x82     0
x84     0
x87     0
x90     0
x92     0
x93     0
x94     0
x97     0
x98     0
x100    0
dtype: int64

### x3
Clean up day names

In [11]:
dfCombine["x3"].unique()

array(['Wed', 'Friday', 'Thursday', 'Tuesday', 'Sunday', 'Saturday',
       'Sat', 'Wednesday', 'Sun', 'Tue', 'Thur', 'Monday', 'Fri', 'Mon'],
      dtype=object)

In [12]:
dfCombine["x3"] = dfCombine["x3"].replace(["Mon"],"Monday")
dfCombine["x3"] = dfCombine["x3"].replace(["Tue"],"Tuesday")
dfCombine["x3"] = dfCombine["x3"].replace(["Wed"],"Wednesday")
dfCombine["x3"] = dfCombine["x3"].replace(["Thur"],"Thursday")
dfCombine["x3"] = dfCombine["x3"].replace(["Fri"],"Friday")
dfCombine["x3"] = dfCombine["x3"].replace(["Sat"],"Saturday")
dfCombine["x3"] = dfCombine["x3"].replace(["Sun"],"Sunday")

In [13]:
dfCombine["x3"].unique()

array(['Wednesday', 'Friday', 'Thursday', 'Tuesday', 'Sunday', 'Saturday',
       'Monday'], dtype=object)

### x7
Remove percent sign

In [14]:
dfCombine['x7'] = dfCombine["x7"].str.replace("%","")

In [15]:
dfCombine.head(5)

y        x1         x2         x3        x4        x5        x6       x7  \
0  0.0  0.165254  18.060003  Wednesday  1.077380 -1.339233 -1.584341   0.0062   
1  1.0  2.441471  18.416307     Friday  1.482586  0.920817 -0.759931   0.0064   
2  1.0  4.427278  19.188092   Thursday  0.145652  0.366093  0.709962   -8e-04   
3  0.0  3.925235  19.901257    Tuesday  1.763602 -0.251926 -0.827461  -0.0057   
4  0.0  2.868802  22.202473     Sunday  3.405119  0.083162  1.381504   0.0109   

         x8        x9       x10       x12       x13       x15       x17  \
0  0.220784  1.816481  1.171788  4.644568  4.814885 -0.587361  4.278444   
1  1.192441  3.513950  1.419900  1.459868  1.443983 -1.165869  8.487132   
2  0.952323  0.782974 -1.247022  1.098525  1.216059  0.211685  8.601193   
3 -0.520756  1.825586  2.223038 -1.390239  3.962961 -2.046856  6.611554   
4 -0.732739  2.151990 -0.275406  7.230125  3.877312 -1.201565  9.103015   

         x18                 x19       x20        x21         x22       x23  \
0  77.709700  $-908.650758424405  7.328029  36.601967  126.384294  1.264713   
1  70.808566   $-1864.9622875143  6.796408   3.789176   93.037021  0.346201   
2  75.922820  $-543.187402955527  3.802395   7.407996   71.022413  0.070968   
3  74.966925  $-182.626380634258  7.728963   8.136213  121.610846  0.700954   
4  77.977156   $967.007090837503  2.272765  24.452102   56.492317 -1.102387   

        x25       x26          x27        x28       x29  x31       x32  \
0  3.834082  2.352406   905.491445  14.779467  0.752524   no -0.597288   
1  4.345028  1.651579   868.784447  25.914680  1.445294   no  1.127376   
2  4.332644 -0.375737  1077.233497  13.177479  4.174536   no  2.605279   
3  7.294990 -0.603983  1051.655489  17.006528  2.347355   no  1.071202   
4  6.313827  0.429187   949.904947  16.962710  0.510985  yes  2.283921   

         x34       x35        x36        x37        x40       x43       x46  \
0  46.667221  3.159002  69.444841  32.450076  18.129499  1.227266  0.234654   
1  47.928569 -1.013712  61.289132  35.521824  11.389081 -0.822055  1.104631   
2  56.674425  0.108486  69.798923  30.684074  14.372443  0.150728 -2.561083   
3  59.154933  1.319711  65.408246  34.401290  13.191173 -0.965711 -2.123048   
4  46.445617  0.022747  66.662910  31.135261  17.210430  1.012093  3.477451   

         x47        x48          x50       x51        x53       x56  \
0  53.359113  58.251700  1602.756464  2.740134  72.167186 -1.532562   
1  37.728834  56.052749  3350.837875  1.995635  73.964560  0.956001   
2  50.236892  63.975108  3756.910196  1.772648  59.428690  4.127857   
3  41.857197  59.226119  1961.609788  3.155214  68.671023  0.663759   
4  47.844153  55.921988  2345.195505  3.253079  71.853326 -2.273628   

          x58  x59        x60       x62       x63       x65       x66  \
0  300.629990    0     August  6.600008  1.252479   farmers  1.212048   
1  300.629990    0      April  3.864254  0.057936  allstate  0.738526   
2  182.369349    0  September  1.370661 -0.239705     geico -0.032980   
3  300.629990    0  September  0.013162  0.318335     geico -0.466535   
4  149.725023    0    January  1.305936  0.418272     geico -0.195764   

         x67        x68       x69       x70       x71        x72       x73  \
0  12.896733  15.263491  0.518653  0.543768 -0.912506  53.521544  7.723719   
1  14.402071  33.940951 -0.140384  1.016703 -0.000234  46.797571  7.260365   
2  14.402071  18.211817 -0.819798  1.010811 -0.000234  48.202036  9.336021   
3  14.402071  14.629914  1.389325  0.704880 -1.510949  49.882647  5.661421   
4  14.402071  12.227512 -2.951163  1.096548 -0.000234  51.349106  9.422401   

        x81       x82       x84       x87       x90       x92  x93       x94  \
0 -0.471594 -3.261865  0.383351 -0.162668  2.770844  0.909479   no  5.492487   
1  1.691428 -4.789183  2.138771  1.252022  3.703013  9.093466   no  3.346429   
2 -2.608974 -6.387984  2.138771  0.657292  3.997183  0.305170   no  4.456565   
3  1.424316 -5.43

In [16]:
# remove "E" notation
dfCombine["x7"] = pd.to_numeric(dfCombine["x7"])
dfCombine.sample(25)

y        x1         x2         x3        x4        x5        x6  \
4010   0.0  2.367744  20.529628   Thursday -0.192869  2.004715 -0.859394   
17304  0.0  2.157367  17.415653   Thursday -0.431223  1.312507 -1.212351   
34087  0.0  5.804372  17.057679  Wednesday -0.695811  0.891840 -1.278062   
10385  1.0  1.686810  20.422436  Wednesday -2.160301  0.365605  1.189986   
49007  0.0  3.365842  21.373347     Monday  1.552885 -2.049815 -0.116837   
6506   1.0  2.703711  17.590990     Friday -0.044775  0.087664  1.777662   
37093  0.0  2.797962  23.005770     Monday  2.115252  3.051937 -1.570434   
31781  0.0  6.152333  19.336132     Sunday  0.379153  1.692065 -0.689577   
28805  0.0  2.704904  18.054402    Tuesday -1.421297  1.791816 -0.169934   
45860  0.0  1.372144  18.684776  Wednesday  0.240627  1.539848 -1.244169   
47567  0.0  4.828767  20.864650  Wednesday  1.641153 -0.833834  4.427113   
36140  0.0  1.668627  17.840394     Friday -0.696464  1.073321  0.094635   
37312  0.0  0.895549  18.114421     Monday -0.010151 -1.386386  0.101300   
20751  0.0  6.334395  18.040835     Friday -0.039931  1.687109 -0.189007   
19889  0.0  2.024065  22.212369     Sunday  1.543326  0.954929  1.239705   
35594  0.0  1.042675  21.099432    Tuesday -1.027704 -0.399396  0.028859   
3368   0.0  2.743285  20.719602     Monday  1.289156  0.241188 -1.931910   
33708  0.0  3.353466  19.656788     Monday -0.741108  0.588387  0.954922   
12166  1.0  2.924649  17.653742   Saturday -0.826307  0.180705 -1.603107   
10871  0.0  2.417404  23.199738  Wednesday -1.899964  0.746305  0.481734   
33080  0.0  5.004815  20.127171     Sunday -0.848632 -0.564679  2.701644   
5311   0.0  3.178404  18.069924    Tuesday -1.919454 -0.705320 -1.376781   
10029  0.0  3.219542  18.308444     Sunday  0.117466 -0.809084 -1.525852   
34072  0.0  4.173465  18.257331  Wednesday  1.404074  0.770137 -1.259545   
26147  0.0  2.166004  20.762058  Wednesday -0.176619  0.267794  0.944848   

           x7        x8        x9       x10       x12       x13       x15  \
4010  -0.0165  1.495263 -0.416752 -0.001031 -0.202942  3.912375  0.594421   
17304 -0.0094  0.373995  4.536326 -1.394070  2.877613  3.259838  2.871844   
34087 -0.0264  0.721206  2.192405 -1.764370  0.326147  6.466948 -1.075984   
10385 -0.0082 -1.300610  3.599191  0.305015  7.630414  5.491448  2.055179   
49007  0.0007  1.648754  3.739607 -0.241030  8.716513  5.000246  2.090907   
6506  -0.0003  0.344275  3.981359  0.813611  6.514089  4.656391  0.915386   
37093 -0.0089 -0.505494  3.149369  0.928416  5.600523  1.641832  3.471249   
31781  0.0021 -0.450959  4.253360  0.654724  7.113846  7.222348 -0.755193   
28805  0.0173 -1.532670  7.827418 -0.243035  8.548759  1.713286  1.620031   
45860 -0.0275  0.642904  5.584981  1.139157  0.543419  6.526048  4.151824   
47567 -0.0028 -0.739831  0.063043  0.814136  1.983971  4.472298 -1.698588   
36140  0.0021  0.156700  2.692333 -2.359079  3.316814  1.460362 -0.848696   
37312  0.0083 -1.511127  1.214507 -0.088522 -0.828917  3.622058 -0.405164   
20751 -0.0052  1.483104  2.778924 -0.412378  4.047160  6.008903  0.577194   
19889  0.0104  0.332072  1.094753 -0.145842 -1.307571  4.427086 -1.577894   
35594 -0.0058  0.250097  2.213913  1.124142  2.215123  2.123005 -0.329619   
3368   0.0129  3.899322  4.103328 -0.007380  5.698074  7.835141 -0.384918   
33708  0.0069  0.527873  2.935190  0.094933  0.438690  1.845398 -1.290537   
12166 -0.0074  3.191851  2.868944  0.168691  4.912333  7.749281  0.219132   
10871  0.0065 -2.376342  1.625268  1.666676 -1.944779  2.849698 -0.342185   
33080  0.0013  0.078601  2.008517  0.701939 -0.219040  1.674062 -0.832737   
5311  -0.0065  0.756643  6.148693  0.118369  6.190754  4.585837 -0.276976   
10029 -0.0061  0.637268  4.100056  0.864635  4.280394  2.036347 -0.600679   
34072  0.0048  1.896090  0.813039  1.284475  5.740335  5.337200 -2.160016   
26147  0.0016 -1.568881  0.988569  0.593339 -0.519581  1.815740 -1.084650   

             x17        x18  

### x19

In [17]:
# replace dollar sign
dfCombine["x19"] = dfCombine["x19"].replace(to_replace="[\$]",value="",regex=True)
dfCombine.head(10)

y        x1         x2         x3        x4        x5        x6      x7  \
0  0.0  0.165254  18.060003  Wednesday  1.077380 -1.339233 -1.584341  0.0062   
1  1.0  2.441471  18.416307     Friday  1.482586  0.920817 -0.759931  0.0064   
2  1.0  4.427278  19.188092   Thursday  0.145652  0.366093  0.709962 -0.0008   
3  0.0  3.925235  19.901257    Tuesday  1.763602 -0.251926 -0.827461 -0.0057   
4  0.0  2.868802  22.202473     Sunday  3.405119  0.083162  1.381504  0.0109   
5  0.0  1.926053  19.524376   Saturday  3.702574  2.300792 -0.331106  0.0079   
6  0.0  0.094162  18.894755   Thursday  2.027918  1.580408  1.190932  0.0174   
7  0.0  0.921300  19.160713   Saturday -1.274067 -0.867023 -2.036134 -0.0106   
8  0.0  0.589354  21.627989  Wednesday  1.672593  0.158446  2.073026  0.0032   
9  0.0  1.016857  21.440000    Tuesday -2.619848  0.637245  1.357868  0.0091   

         x8        x9       x10       x12       x13       x15       x17  \
0  0.220784  1.816481  1.171788  4.644568  4.814885 -0.587361  4.278444   
1  1.192441  3.513950  1.419900  1.459868  1.443983 -1.165869  8.487132   
2  0.952323  0.782974 -1.247022  1.098525  1.216059  0.211685  8.601193   
3 -0.520756  1.825586  2.223038 -1.390239  3.962961 -2.046856  6.611554   
4 -0.732739  2.151990 -0.275406  7.230125  3.877312 -1.201565  9.103015   
5  3.085443  0.472701  1.943171 -3.582885  3.522031 -2.799700  7.283226   
6 -0.331730  2.178622 -0.454172  1.762009  2.125462 -0.907832  8.631917   
7  2.830838  1.206701  1.315474 -2.483859  4.224671 -2.123287  6.667980   
8 -0.908026 -0.064304  0.861625 -2.124964  1.903264 -3.364331  5.437052   
9  1.471836  0.811922 -0.699837  2.893282  1.603635 -3.283478  7.988783   

         x18                x19        x20        x21         x22       x23  \
0  77.709700  -908.650758424405   7.328029  36.601967  126.384294  1.264713   
1  70.808566   -1864.9622875143   6.796408   3.789176   93.037021  0.346201   
2  75.922820  -543.187402955527   3.802395   7.407996   71.022413  0.070968   
3  74.966925  -182.626380634258   7.728963   8.136213  121.610846  0.700954   
4  77.977156   967.007090837503   2.272765  24.452102   56.492317 -1.102387   
5  87.734440   271.163475698959  15.770180  44.874132   45.729121  0.320526   
6  71.282134   749.965009681221  14.220882  12.727490   40.145649 -0.196627   
7  78.224602   32.7532387681275  10.688299   4.209199   55.802747  1.255861   
8  81.846939   1079.84051477389   8.314540   1.074290   71.459845  1.175947   
9  89.462640  -1668.77796649785   4.055610  -0.328695   24.882245 -0.102272   

        x25       x26          x27        x28       x29  x31       x32  \
0  3.834082  2.352406   905.491445  14.779467  0.752524   no -0.597288   
1  4.345028  1.651579   868.784447  25.914680  1.445294   no  1.127376   
2  4.332644 -0.375737  1077.233497  13.177479  4.174536   no  2.605279   
3  7.294990 -0.603983  1051.655489  17.006528  2.347355   no  1.071202   
4  6.313827  0.429187   949.904947  16.962710  0.510985  yes  2.283921   
5  2.245217  4.303686  1187.333248   8.139008  2.233107   no  0.488196   
6  4.784640  1.051441   902.058622  17.881615  1.580413   no  2.424133   
7  3.560273  1.048213   990.811632  16.443997  0.241411  yes  0.123102   
8  2.640861 -0.255633  1214.321482  13.353094 -0.060317   no  0.368641   
9  2.166865 -0.795430  1244.523235   9.513302  0.273440   no  0.007047   

         x34       x35        x36        x37        x40       x43       x46  \
0  46.667221  3.159002  69.444841  32.450076  18.129499  1.227266  0.234654   
1  47.928569 -1.013712  61.289132  35.521824  11.389081 -0.822055  1.104631   
2  56.674425  0.108486  69.798923  30.684074  14.372443  0.150728 -2.561083   
3  59.154933  1.319711  65.408246  34.401290  13.191173 -0.965711 -2.123048   
4  46.445617  0.022747  66.662910  31.135261  17.210430  1.012093  3.477451   
5  45.559596  0.359074  76.835865  37.602918  12.628248  1.452704 -1.573853   
6  53.786798  2.610824  68.051870  32.855520  12.243767  1.8

### Round to 6 decimal places

In [18]:
dfCombine["x19"] = dfCombine["x19"].astype(float).round(6)
dfCombine.head(10)

y        x1         x2         x3        x4        x5        x6      x7  \
0  0.0  0.165254  18.060003  Wednesday  1.077380 -1.339233 -1.584341  0.0062   
1  1.0  2.441471  18.416307     Friday  1.482586  0.920817 -0.759931  0.0064   
2  1.0  4.427278  19.188092   Thursday  0.145652  0.366093  0.709962 -0.0008   
3  0.0  3.925235  19.901257    Tuesday  1.763602 -0.251926 -0.827461 -0.0057   
4  0.0  2.868802  22.202473     Sunday  3.405119  0.083162  1.381504  0.0109   
5  0.0  1.926053  19.524376   Saturday  3.702574  2.300792 -0.331106  0.0079   
6  0.0  0.094162  18.894755   Thursday  2.027918  1.580408  1.190932  0.0174   
7  0.0  0.921300  19.160713   Saturday -1.274067 -0.867023 -2.036134 -0.0106   
8  0.0  0.589354  21.627989  Wednesday  1.672593  0.158446  2.073026  0.0032   
9  0.0  1.016857  21.440000    Tuesday -2.619848  0.637245  1.357868  0.0091   

         x8        x9       x10       x12       x13       x15       x17  \
0  0.220784  1.816481  1.171788  4.644568  4.814885 -0.587361  4.278444   
1  1.192441  3.513950  1.419900  1.459868  1.443983 -1.165869  8.487132   
2  0.952323  0.782974 -1.247022  1.098525  1.216059  0.211685  8.601193   
3 -0.520756  1.825586  2.223038 -1.390239  3.962961 -2.046856  6.611554   
4 -0.732739  2.151990 -0.275406  7.230125  3.877312 -1.201565  9.103015   
5  3.085443  0.472701  1.943171 -3.582885  3.522031 -2.799700  7.283226   
6 -0.331730  2.178622 -0.454172  1.762009  2.125462 -0.907832  8.631917   
7  2.830838  1.206701  1.315474 -2.483859  4.224671 -2.123287  6.667980   
8 -0.908026 -0.064304  0.861625 -2.124964  1.903264 -3.364331  5.437052   
9  1.471836  0.811922 -0.699837  2.893282  1.603635 -3.283478  7.988783   

         x18          x19        x20        x21         x22       x23  \
0  77.709700  -908.650758   7.328029  36.601967  126.384294  1.264713   
1  70.808566 -1864.962288   6.796408   3.789176   93.037021  0.346201   
2  75.922820  -543.187403   3.802395   7.407996   71.022413  0.070968   
3  74.966925  -182.626381   7.728963   8.136213  121.610846  0.700954   
4  77.977156   967.007091   2.272765  24.452102   56.492317 -1.102387   
5  87.734440   271.163476  15.770180  44.874132   45.729121  0.320526   
6  71.282134   749.965010  14.220882  12.727490   40.145649 -0.196627   
7  78.224602    32.753239  10.688299   4.209199   55.802747  1.255861   
8  81.846939  1079.840515   8.314540   1.074290   71.459845  1.175947   
9  89.462640 -1668.777966   4.055610  -0.328695   24.882245 -0.102272   

        x25       x26          x27        x28       x29  x31       x32  \
0  3.834082  2.352406   905.491445  14.779467  0.752524   no -0.597288   
1  4.345028  1.651579   868.784447  25.914680  1.445294   no  1.127376   
2  4.332644 -0.375737  1077.233497  13.177479  4.174536   no  2.605279   
3  7.294990 -0.603983  1051.655489  17.006528  2.347355   no  1.071202   
4  6.313827  0.429187   949.904947  16.962710  0.510985  yes  2.283921   
5  2.245217  4.303686  1187.333248   8.139008  2.233107   no  0.488196   
6  4.784640  1.051441   902.058622  17.881615  1.580413   no  2.424133   
7  3.560273  1.048213   990.811632  16.443997  0.241411  yes  0.123102   
8  2.640861 -0.255633  1214.321482  13.353094 -0.060317   no  0.368641   
9  2.166865 -0.795430  1244.523235   9.513302  0.273440   no  0.007047   

         x34       x35        x36        x37        x40       x43       x46  \
0  46.667221  3.159002  69.444841  32.450076  18.129499  1.227266  0.234654   
1  47.928569 -1.013712  61.289132  35.521824  11.389081 -0.822055  1.104631   
2  56.674425  0.108486  69.798923  30.684074  14.372443  0.150728 -2.561083   
3  59.154933  1.319711  65.408246  34.401290  13.191173 -0.965711 -2.123048   
4  46.445617  0.022747  66.662910  31.135261  17.210430  1.012093  3.477451   
5  45.559596  0.359074  76.835865  37.602918  12.628248  1.452704 -1.573853   
6  53.786798  2.610824  68.051870  32.855520  12.243767  1.863568  0.977259   
7  54.209775  2.215420  70.230786  31.665864  1

### x31

In [19]:
# map 0 to no and 1 to yes
dfCombine["x31"] = dfCombine["x31"].map({"no":0,"yes":1})
dfCombine["x31"].unique()

array([0, 1], dtype=int64)

### x93

In [20]:
dfCombine["x93"] = dfCombine["x93"].map({"no":0,"yes":1})
dfCombine["x93"].unique()

array([0, 1], dtype=int64)

### Encode categorical variables

### x3

In [21]:
# days of the week
x3 = pd.get_dummies(dfCombine['x3'],prefix='cat_')
dfCombine = pd.merge(
    left=dfCombine,
    right=x3,
    left_index=True,
    right_index=True,
)

dfCombine.head(15)

y        x1         x2         x3        x4        x5        x6      x7  \
0   0.0  0.165254  18.060003  Wednesday  1.077380 -1.339233 -1.584341  0.0062   
1   1.0  2.441471  18.416307     Friday  1.482586  0.920817 -0.759931  0.0064   
2   1.0  4.427278  19.188092   Thursday  0.145652  0.366093  0.709962 -0.0008   
3   0.0  3.925235  19.901257    Tuesday  1.763602 -0.251926 -0.827461 -0.0057   
4   0.0  2.868802  22.202473     Sunday  3.405119  0.083162  1.381504  0.0109   
5   0.0  1.926053  19.524376   Saturday  3.702574  2.300792 -0.331106  0.0079   
6   0.0  0.094162  18.894755   Thursday  2.027918  1.580408  1.190932  0.0174   
7   0.0  0.921300  19.160713   Saturday -1.274067 -0.867023 -2.036134 -0.0106   
8   0.0  0.589354  21.627989  Wednesday  1.672593  0.158446  2.073026  0.0032   
9   0.0  1.016857  21.440000    Tuesday -2.619848  0.637245  1.357868  0.0091   
10  1.0  0.853152  20.300924  Wednesday -0.930155 -0.607742 -0.289608 -0.0052   
11  0.0  4.304882  19.302074     Sunday  2.470847 -1.852730  0.400993 -0.0025   
12  0.0  1.799370  20.725235     Friday -1.381245  0.716132  0.503553 -0.0045   
13  0.0  3.671553  22.207880    Tuesday  0.698724 -1.862047  2.153832  0.0068   
14  0.0  3.567967  16.644123  Wednesday -0.951598 -1.404246 -2.013598 -0.0137   

          x8        x9       x10       x12       x13       x15        x17  \
0   0.220784  1.816481  1.171788  4.644568  4.814885 -0.587361   4.278444   
1   1.192441  3.513950  1.419900  1.459868  1.443983 -1.165869   8.487132   
2   0.952323  0.782974 -1.247022  1.098525  1.216059  0.211685   8.601193   
3  -0.520756  1.825586  2.223038 -1.390239  3.962961 -2.046856   6.611554   
4  -0.732739  2.151990 -0.275406  7.230125  3.877312 -1.201565   9.103015   
5   3.085443  0.472701  1.943171 -3.582885  3.522031 -2.799700   7.283226   
6  -0.331730  2.178622 -0.454172  1.762009  2.125462 -0.907832   8.631917   
7   2.830838  1.206701  1.315474 -2.483859  4.224671 -2.123287   6.667980   
8  -0.908026 -0.064304  0.861625 -2.124964  1.903264 -3.364331   5.437052   
9   1.471836  0.811922 -0.699837  2.893282  1.603635 -3.283478   7.988783   
10 -0.040689  1.858641  1.491314 -1.600497  3.219078 -0.764358   8.769019   
11  0.907410  2.052139  1.186982  3.919136  4.463684 -2.649127   8.539390   
12  0.410077 -0.186681 -1.054522  1.021625  6.562940  0.201709  10.347098   
13  0.205453  1.309530  0.683070  1.433037  2.625811 -0.824623   6.134170   
14  1.578542  1.905247 -1.224807  1.194875  1.661333 -2.137908   3.995859   

          x18          x19        x20        x21         x22       x23  \
0   77.709700  -908.650758   7.328029  36.601967  126.384294  1.264713   
1   70.808566 -1864.962288   6.796408   3.789176   93.037021  0.346201   
2   75.922820  -543.187403   3.802395   7.407996   71.022413  0.070968   
3   74.966925  -182.626381   7.728963   8.136213  121.610846  0.700954   
4   77.977156   967.007091   2.272765  24.452102   56.492317 -1.102387   
5   87.734440   271.163476  15.770180  44.874132   45.729121  0.320526   
6   71.282134   749.965010  14.220882  12.727490   40.145649 -0.196627   
7   78.224602    32.753239  10.688299   4.209199   55.802747  1.255861   
8   81.846939  1079.840515   8.314540   1.074290   71.459845  1.175947   
9   89.462640 -1668.777966   4.055610  -0.328695   24.882245 -0.102272   
10  82.870050  1391.302605   7.099256   6.247910   59.310311 -1.107675   
11  78.858807 -1822.863099   7.000926  19.971813   95.353723 -1.060869   
12  69.524087  1038.634263   6.341097   0.817180   97.972638  1.783286   
13  72.430917  -377.495529   7.074143   7.676579   72.703358  0.520186   
14  89.192758   581.353320  10.061368   6.925499   71.816502 -0.724962   

         x25       x26          x27        x28       x29  x31       x32  \
0   3.834082  2.352406   905.491445  14.779467  0.752524    0 -0.597288   
1   4.345028  1.651579   868.784447  25.914680  1.445294    0  1.127376   
2   4.332644 -0.375737  1077.233497  13.177479  4.174536    0  2.6

#### x60

In [22]:
# months
x60 = pd.get_dummies(dfCombine['x60'],prefix='cat_')
dfCombine = pd.merge(
    left=dfCombine,
    right=x60,
    left_index=True,
    right_index=True,
)

dfCombine.head(15)

y        x1         x2         x3        x4        x5        x6      x7  \
0   0.0  0.165254  18.060003  Wednesday  1.077380 -1.339233 -1.584341  0.0062   
1   1.0  2.441471  18.416307     Friday  1.482586  0.920817 -0.759931  0.0064   
2   1.0  4.427278  19.188092   Thursday  0.145652  0.366093  0.709962 -0.0008   
3   0.0  3.925235  19.901257    Tuesday  1.763602 -0.251926 -0.827461 -0.0057   
4   0.0  2.868802  22.202473     Sunday  3.405119  0.083162  1.381504  0.0109   
5   0.0  1.926053  19.524376   Saturday  3.702574  2.300792 -0.331106  0.0079   
6   0.0  0.094162  18.894755   Thursday  2.027918  1.580408  1.190932  0.0174   
7   0.0  0.921300  19.160713   Saturday -1.274067 -0.867023 -2.036134 -0.0106   
8   0.0  0.589354  21.627989  Wednesday  1.672593  0.158446  2.073026  0.0032   
9   0.0  1.016857  21.440000    Tuesday -2.619848  0.637245  1.357868  0.0091   
10  1.0  0.853152  20.300924  Wednesday -0.930155 -0.607742 -0.289608 -0.0052   
11  0.0  4.304882  19.302074     Sunday  2.470847 -1.852730  0.400993 -0.0025   
12  0.0  1.799370  20.725235     Friday -1.381245  0.716132  0.503553 -0.0045   
13  0.0  3.671553  22.207880    Tuesday  0.698724 -1.862047  2.153832  0.0068   
14  0.0  3.567967  16.644123  Wednesday -0.951598 -1.404246 -2.013598 -0.0137   

          x8        x9       x10       x12       x13       x15        x17  \
0   0.220784  1.816481  1.171788  4.644568  4.814885 -0.587361   4.278444   
1   1.192441  3.513950  1.419900  1.459868  1.443983 -1.165869   8.487132   
2   0.952323  0.782974 -1.247022  1.098525  1.216059  0.211685   8.601193   
3  -0.520756  1.825586  2.223038 -1.390239  3.962961 -2.046856   6.611554   
4  -0.732739  2.151990 -0.275406  7.230125  3.877312 -1.201565   9.103015   
5   3.085443  0.472701  1.943171 -3.582885  3.522031 -2.799700   7.283226   
6  -0.331730  2.178622 -0.454172  1.762009  2.125462 -0.907832   8.631917   
7   2.830838  1.206701  1.315474 -2.483859  4.224671 -2.123287   6.667980   
8  -0.908026 -0.064304  0.861625 -2.124964  1.903264 -3.364331   5.437052   
9   1.471836  0.811922 -0.699837  2.893282  1.603635 -3.283478   7.988783   
10 -0.040689  1.858641  1.491314 -1.600497  3.219078 -0.764358   8.769019   
11  0.907410  2.052139  1.186982  3.919136  4.463684 -2.649127   8.539390   
12  0.410077 -0.186681 -1.054522  1.021625  6.562940  0.201709  10.347098   
13  0.205453  1.309530  0.683070  1.433037  2.625811 -0.824623   6.134170   
14  1.578542  1.905247 -1.224807  1.194875  1.661333 -2.137908   3.995859   

          x18          x19        x20        x21         x22       x23  \
0   77.709700  -908.650758   7.328029  36.601967  126.384294  1.264713   
1   70.808566 -1864.962288   6.796408   3.789176   93.037021  0.346201   
2   75.922820  -543.187403   3.802395   7.407996   71.022413  0.070968   
3   74.966925  -182.626381   7.728963   8.136213  121.610846  0.700954   
4   77.977156   967.007091   2.272765  24.452102   56.492317 -1.102387   
5   87.734440   271.163476  15.770180  44.874132   45.729121  0.320526   
6   71.282134   749.965010  14.220882  12.727490   40.145649 -0.196627   
7   78.224602    32.753239  10.688299   4.209199   55.802747  1.255861   
8   81.846939  1079.840515   8.314540   1.074290   71.459845  1.175947   
9   89.462640 -1668.777966   4.055610  -0.328695   24.882245 -0.102272   
10  82.870050  1391.302605   7.099256   6.247910   59.310311 -1.107675   
11  78.858807 -1822.863099   7.000926  19.971813   95.353723 -1.060869   
12  69.524087  1038.634263   6.341097   0.817180   97.972638  1.783286   
13  72.430917  -377.495529   7.074143   7.676579   72.703358  0.520186   
14  89.192758   581.353320  10.061368   6.925499   71.816502 -0.724962   

         x25       x26          x27        x28       x29  x31       x32  \
0   3.834082  2.352406   905.491445  14.779467  0.752524    0 -0.597288   
1   4.345028  1.651579   868.784447  25.914680  1.445294    0  1.127376   
2   4.332644 -0.375737  1077.233497  13.177479  4.174536    0  2.6

#### x65

In [23]:
# insurance name
x65 = pd.get_dummies(dfCombine['x65'],prefix='cat_')
dfCombine = pd.merge(
    left=dfCombine,
    right=x65,
    left_index=True,
    right_index=True,
)

dfCombine.head(15)

y        x1         x2         x3        x4        x5        x6      x7  \
0   0.0  0.165254  18.060003  Wednesday  1.077380 -1.339233 -1.584341  0.0062   
1   1.0  2.441471  18.416307     Friday  1.482586  0.920817 -0.759931  0.0064   
2   1.0  4.427278  19.188092   Thursday  0.145652  0.366093  0.709962 -0.0008   
3   0.0  3.925235  19.901257    Tuesday  1.763602 -0.251926 -0.827461 -0.0057   
4   0.0  2.868802  22.202473     Sunday  3.405119  0.083162  1.381504  0.0109   
5   0.0  1.926053  19.524376   Saturday  3.702574  2.300792 -0.331106  0.0079   
6   0.0  0.094162  18.894755   Thursday  2.027918  1.580408  1.190932  0.0174   
7   0.0  0.921300  19.160713   Saturday -1.274067 -0.867023 -2.036134 -0.0106   
8   0.0  0.589354  21.627989  Wednesday  1.672593  0.158446  2.073026  0.0032   
9   0.0  1.016857  21.440000    Tuesday -2.619848  0.637245  1.357868  0.0091   
10  1.0  0.853152  20.300924  Wednesday -0.930155 -0.607742 -0.289608 -0.0052   
11  0.0  4.304882  19.302074     Sunday  2.470847 -1.852730  0.400993 -0.0025   
12  0.0  1.799370  20.725235     Friday -1.381245  0.716132  0.503553 -0.0045   
13  0.0  3.671553  22.207880    Tuesday  0.698724 -1.862047  2.153832  0.0068   
14  0.0  3.567967  16.644123  Wednesday -0.951598 -1.404246 -2.013598 -0.0137   

          x8        x9       x10       x12       x13       x15        x17  \
0   0.220784  1.816481  1.171788  4.644568  4.814885 -0.587361   4.278444   
1   1.192441  3.513950  1.419900  1.459868  1.443983 -1.165869   8.487132   
2   0.952323  0.782974 -1.247022  1.098525  1.216059  0.211685   8.601193   
3  -0.520756  1.825586  2.223038 -1.390239  3.962961 -2.046856   6.611554   
4  -0.732739  2.151990 -0.275406  7.230125  3.877312 -1.201565   9.103015   
5   3.085443  0.472701  1.943171 -3.582885  3.522031 -2.799700   7.283226   
6  -0.331730  2.178622 -0.454172  1.762009  2.125462 -0.907832   8.631917   
7   2.830838  1.206701  1.315474 -2.483859  4.224671 -2.123287   6.667980   
8  -0.908026 -0.064304  0.861625 -2.124964  1.903264 -3.364331   5.437052   
9   1.471836  0.811922 -0.699837  2.893282  1.603635 -3.283478   7.988783   
10 -0.040689  1.858641  1.491314 -1.600497  3.219078 -0.764358   8.769019   
11  0.907410  2.052139  1.186982  3.919136  4.463684 -2.649127   8.539390   
12  0.410077 -0.186681 -1.054522  1.021625  6.562940  0.201709  10.347098   
13  0.205453  1.309530  0.683070  1.433037  2.625811 -0.824623   6.134170   
14  1.578542  1.905247 -1.224807  1.194875  1.661333 -2.137908   3.995859   

          x18          x19        x20        x21         x22       x23  \
0   77.709700  -908.650758   7.328029  36.601967  126.384294  1.264713   
1   70.808566 -1864.962288   6.796408   3.789176   93.037021  0.346201   
2   75.922820  -543.187403   3.802395   7.407996   71.022413  0.070968   
3   74.966925  -182.626381   7.728963   8.136213  121.610846  0.700954   
4   77.977156   967.007091   2.272765  24.452102   56.492317 -1.102387   
5   87.734440   271.163476  15.770180  44.874132   45.729121  0.320526   
6   71.282134   749.965010  14.220882  12.727490   40.145649 -0.196627   
7   78.224602    32.753239  10.688299   4.209199   55.802747  1.255861   
8   81.846939  1079.840515   8.314540   1.074290   71.459845  1.175947   
9   89.462640 -1668.777966   4.055610  -0.328695   24.882245 -0.102272   
10  82.870050  1391.302605   7.099256   6.247910   59.310311 -1.107675   
11  78.858807 -1822.863099   7.000926  19.971813   95.353723 -1.060869   
12  69.524087  1038.634263   6.341097   0.817180   97.972638  1.783286   
13  72.430917  -377.495529   7.074143   7.676579   72.703358  0.520186   
14  89.192758   581.353320  10.061368   6.925499   71.816502 -0.724962   

         x25       x26          x27        x28       x29  x31       x32  \
0   3.834082  2.352406   905.491445  14.779467  0.752524    0 -0.597288   
1   4.345028  1.651579   868.784447  25.914680  1.445294    0  1.127376   
2   4.332644 -0.375737  1077.233497  13.177479  4.174536    0  2.6

In [24]:
dfTrainClean = dfCombine.iloc[:40000,]
dfTestClean = dfCombine.iloc[40000:,]

### Logistic regression (classification) & neural network

class sklearn.linear_model.LogisticRegression(penalty='l2',dual=False,tol=0.0001,C=1.0,fit_intercept=True,
intercept_scaling=1,class_weight=None,random_state=None,solver='lbfgs',max_iter=100,multi_class='auto', 
verbose=0,warm_start=False,n_jobs=None,l1_ratio=None)

class sklearn.neural_network.MLPClassifier(hidden_layer_sizes=(100, ), activation='relu', *, 
solver='adam',alpha=0.0001,batch_size='auto', learning_rate='constant', learning_rate_init=0.001,
power_t=0.5,max_iter=200,shuffle=True, random_state=None, tol=0.0001, verbose=False, 
warm_start=False, momentum=0.9,nesterovs_momentum=True, early_stopping=False, 
validation_fraction=0.1,beta_1=0.9,beta_2=0.999,epsilon=1e-08,n_iter_no_change=10, max_fun=15000)

In [25]:
dfCombine.columns

Index(['y', 'x1', 'x2', 'x3', 'x4', 'x5', 'x6', 'x7', 'x8', 'x9', 'x10', 'x12',
       'x13', 'x15', 'x17', 'x18', 'x19', 'x20', 'x21', 'x22', 'x23', 'x25',
       'x26', 'x27', 'x28', 'x29', 'x31', 'x32', 'x34', 'x35', 'x36', 'x37',
       'x40', 'x43', 'x46', 'x47', 'x48', 'x50', 'x51', 'x53', 'x56', 'x58',
       'x59', 'x60', 'x62', 'x63', 'x65', 'x66', 'x67', 'x68', 'x69', 'x70',
       'x71', 'x72', 'x73', 'x81', 'x82', 'x84', 'x87', 'x90', 'x92', 'x93',
       'x94', 'x97', 'x98', 'x100', 'cat__Friday', 'cat__Monday',
       'cat__Saturday', 'cat__Sunday', 'cat__Thursday', 'cat__Tuesday',
       'cat__Wednesday', 'cat__April', 'cat__August', 'cat__December',
       'cat__February', 'cat__January', 'cat__July', 'cat__June', 'cat__March',
       'cat__May', 'cat__November', 'cat__October', 'cat__September',
       'cat__allstate', 'cat__esurance', 'cat__farmers', 'cat__geico',
       'cat__progressive'],
      dtype='object')

In [26]:
features = [
    'x1','x2','x4','x5','x6','x7','x8','x9','x10','x12',
    'x13','x15','x17','x18', 'x19', 'x20', 'x21','x22', 'x23', 'x25',
    'x26','x27','x28','x29', 'x31', 'x32', 'x34','x35', 'x36', 'x37',
    'x40','x43','x46','x47', 'x48', 'x50', 'x51','x53', 'x56', 'x58',
    'x59','x62','x63','x66', 'x67', 'x68', 'x69','x70',
    'x71','x72','x73','x81', 'x82', 'x84', 'x87','x90','x92','x93',
    'x94','x97','x98', 'x100', 'cat__Friday', 'cat__Monday',
    'cat__Saturday','cat__Sunday','cat__Thursday', 'cat__Tuesday',
    'cat__Wednesday','cat__April','cat__August','cat__December',
    'cat__February','cat__January','cat__July','cat__June', 'cat__March',
    'cat__May','cat__November', 'cat__October', 'cat__September',
    'cat__allstate','cat__esurance', 'cat__farmers', 'cat__geico',
    'cat__progressive']

# x4 x6 x9 x10 x12 x17 x22 x23 x34 x48 x53 x59 x69 x73 x81 x92 x97
dropFeatures = [
    'x1','x2','x5','x7','x8',
    'x13','x15','x18','x19', 'x20','x21','x25',
    'x26','x27','x28','x29', 'x31','x32','x35', 'x36', 'x37',
    'x40','x43','x46','x47', 'x50','x51', 'x56','x58',
    'x62','x63','x66', 'x67', 'x68','x70',
    'x71','x72','x82','x84', 'x87','x90','x93',
    'x94','x98', 'x100', 'cat__Friday', 'cat__Monday',
    'cat__Saturday','cat__Sunday','cat__Thursday', 'cat__Tuesday',
    'cat__Wednesday','cat__April','cat__August','cat__December',
    'cat__February','cat__January','cat__July','cat__June', 'cat__March',
    'cat__May','cat__November', 'cat__October', 'cat__September',
    'cat__allstate','cat__esurance', 'cat__farmers', 'cat__geico',
    'cat__progressive']

### Scale data

In [27]:
scalerX = StandardScaler()
scalerX.fit(dfTrainClean[dropFeatures])
xTrainScaled = scalerX.transform(dfTrainClean[dropFeatures])
xTestScaled = scalerX.transform(dfTestClean[dropFeatures])

In [28]:
# logistic regression
logReg = LogisticRegression(C=7.7777,solver="lbfgs",max_iter=1111,tol=0.000003333, \
    multi_class="ovr",class_weight=None)

logReg.fit(xTrainScaled.astype(float),dfTrainClean["y"].astype(int))

LogisticRegression(C=7.7777, max_iter=1111, multi_class='ovr', tol=3.333e-06)

In [29]:
# neural networks
mlp = MLPClassifier(hidden_layer_sizes=(76,76,),random_state=None,alpha=0.1111, \
    solver='sgd',learning_rate="adaptive",activation="relu",n_iter_no_change=10, \
    tol=0.00001111,learning_rate_init=0.001111,max_iter=253,momentum=0.9111, \
    batch_size="auto",shuffle=True,warm_start=False,early_stopping=False,validation_fraction=0.1)

mlp.fit(xTrainScaled.astype(float),dfTrainClean["y"].astype(int))

C:\Users\TamLe\anaconda3\envs\sports\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:614: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (253) reached and the optimization hasn't converged yet.
  warnings.warn(


MLPClassifier(alpha=0.1111, hidden_layer_sizes=(76, 76),
              learning_rate='adaptive', learning_rate_init=0.001111,
              max_iter=253, momentum=0.9111, solver='sgd', tol=1.111e-05)

### Predictions

In [30]:
yPredTrainLog = logReg.predict(xTrainScaled.astype(float))
yPredTestLog = logReg.predict(xTestScaled.astype(float))
yProbTrainLog = logReg.predict_proba(xTrainScaled.astype(float))
yProbTestLog = logReg.predict_proba(xTestScaled.astype(float))

In [43]:
yPredTrainNN = mlp.predict(xTrainScaled.astype(float))
yPredTestNN = mlp.predict(xTestScaled.astype(float))
yProbTrainNN = mlp.predict_proba(xTrainScaled.astype(float))
yProbTestNN = mlp.predict_proba(xTestScaled.astype(float))

In [44]:
pd.DataFrame(yPredTrainLog).value_counts()

0    39238
1      762
dtype: int64

### Score

In [34]:
# logistic regression
logisticAccuracy = accuracy_score(dfTrainClean["y"].astype(int),yPredTrainLog.astype(int))
print("accuracy of training set: %s" % accuracy)


accuracy of training set: 0.858725


In [46]:
# neural networks
neuralNetworksAccuracy = accuracy_score(dfTrainClean["y"].astype(int),yPredTrainNN.astype(int))
print("accuracy of training set: %s" % neuralNetworksAccuracy)

accuracy of training set: 0.871125


In [47]:
pd.DataFrame(dfTrainClean["y"]).value_counts()

y  
0.0    34197
1.0     5803
dtype: int64

In [38]:
pd.DataFrame(yPredTrainLog).value_counts()

0    39238
1      762
dtype: int64

In [48]:
pd.DataFrame(yPredTrainNN).value_counts()

0    38414
1     1586
dtype: int64

### Probabilities on training set

In [49]:
pd.DataFrame(yProbTrainLog).head(20)

0         1
0   0.959971  0.040029
1   0.658801  0.341199
2   0.828818  0.171182
3   0.549976  0.450024
4   0.983968  0.016032
5   0.888512  0.111488
6   0.958273  0.041727
7   0.859887  0.140113
8   0.924804  0.075196
9   0.854813  0.145187
10  0.703103  0.296897
11  0.647305  0.352695
12  0.953554  0.046446
13  0.959665  0.040335
14  0.962788  0.037212
15  0.643323  0.356677
16  0.904401  0.095599
17  0.789453  0.210547
18  0.980343  0.019657
19  0.698038  0.301962

In [50]:
pd.DataFrame(yProbTestLog).tail(25)

0         1
9975  0.946557  0.053443
9976  0.909108  0.090892
9977  0.808916  0.191084
9978  0.656960  0.343040
9979  0.330718  0.669282
9980  0.838664  0.161336
9981  0.948758  0.051242
9982  0.948408  0.051592
9983  0.943232  0.056768
9984  0.768390  0.231610
9985  0.799879  0.200121
9986  0.727411  0.272589
9987  0.768810  0.231190
9988  0.941240  0.058760
9989  0.763365  0.236635
9990  0.957531  0.042469
9991  0.951890  0.048110
9992  0.910506  0.089494
9993  0.908765  0.091235
9994  0.932012  0.067988
9995  0.971479  0.028521
9996  0.731335  0.268665
9997  0.898240  0.101760
9998  0.780030  0.219970
9999  0.971902  0.028098

In [51]:
pd.DataFrame(yProbTestLog).iloc[:,[1]].head(15)

1
0   0.020053
1   0.027552
2   0.259752
3   0.177349
4   0.013140
5   0.057507
6   0.274311
7   0.059734
8   0.326363
9   0.623363
10  0.224240
11  0.406556
12  0.677551
13  0.187008
14  0.188720

### Save results

In [52]:
pd.DataFrame(yProbTestLog).iloc[:,[1]].to_csv("glmresults.csv",header=False,index=False)
pd.DataFrame(yProbTestNN).iloc[:,[1]].to_csv("nonglmresults.csv",header=False,index=False)

In [53]:
print(len(yProbTestNN))

10000


### Statsmodels

In [58]:
xInclConstant = sm.add_constant(dfTrainClean[dropFeatures])
model = sm.Logit(dfTrainClean["y"].astype(int),xInclConstant.astype(float))
# logistic regression
results = model.fit()
results.summary()

Optimization terminated successfully.
         Current function value: 0.359918
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                      y   No. Observations:                40000
Model:                          Logit   Df Residuals:                    39933
Method:                           MLE   Df Model:                           66
Date:                Fri, 11 Jun 2021   Pseudo R-squ.:                  0.1308
Time:                        09:52:49   Log-Likelihood:                -14397.
converged:                       True   LL-Null:                       -16563.
Covariance Type:            nonrobust   LLR p-value:                     0.000
====================================================================================
                       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------
const                3.5752   2.06e+06   1.73e-06      1.000   -4.04e+06    4.04e+06
x1                   0.0259      0.008      3.294      0.001       0.011       0.041
x2                   0.0954      0.022      4.407      0.000       0.053       0.138
x5                   0.0446      0.016      2.826      0.005       0.014       0.076
x7                 -74.1062      1.636    -45.290      0.000     -77.313     -70.899
x8                   0.1096      0.018      6.179      0.000       0.075       0.144
x13                  0.0194      0.009      2.108      0.035       0.001       0.037
x15                 -0.0205      0.014     -1.494      0.135      -0.047       0.006
x18                 -0.0094      0.005     -2.034      0.042      -0.018      -0.000
x19                 -0.0001   1.51e-05     -9.304      0.000      -0.000      -0.000
x20                 -0.0301      0.006     -4.903      0.000      -0.042      -0.018
x21                  0.0012      0.001      1.199      0.231      -0.001       0.003
x25                  0.0058      0.006      0.910      0.363      -0.007       0.018
x26                  0.0143      0.014      1.050      0.294      -0.012       0.041
x27                 -0.0004      0.000     -1.429      0.153      -0.001       0.000
x28                 -0.0973      0.012     -7.867      0.000      -0.122      -0.073
x29                 -0.0202      0.013     -1.574      0.115      -0.045       0.005
x31                 -0.5720      0.059     -9.740      0.000      -0.687      -0.457
x32                  0.0144      0.009      1.642      0.101      -0.003       0.032
x35                  0.0243      0.015      1.670      0.095      -0.004       0.053
x36                 -0.0115      0.010     -1.209      0.227      -0.030       0.007
x37                  0.0623      0.009      6.603      0.000       0.044       0.081
x40                 -0.1068      0.008    -13.138      0.000      -0.123      -0.091
x43                 -0.0168      0.015     -1.118      0.264      -0.046       0.013
x46                 -0.0549      0.012     -4.517      0.000      -0.079      -0.031
x47                 -0.1394      0.008    -16.839      0.000      -0.156      -0.123
x50              -9.406e-05   2.48e-05     -3.796      0.000      -0.000   -4.55e-05
x51                  0.0436      0.010      4.290      0.000       0.024       0.063
x56                 -0.0230      0.011     -2.158      0.031      -0.044      -0.002
x58                  0.0007      0.000      1.596      0.110      -0.000       0.002
x62                 -0.0533      0.013     -3.975      0.000      -0.080      -0.027
x63                  0.0444      0.013      3.485      0.000       0.019       0.069
x66                  0.0689      0.016      4.193      0.000       0.037       0.101
x67                 -0.0254      0.014     -1.867      0.062      -0.052       0.001
x68                  0.0145      0.003      4.265      0.000       0.008       0.021
x70        

In [61]:
pd.DataFrame(results.predict()).head(25)

0
0   0.040019
1   0.341241
2   0.171202
3   0.450079
4   0.016030
5   0.111500
6   0.041731
7   0.140101
8   0.075196
9   0.145193
10  0.296935
11  0.352792
12  0.046438
13  0.040331
14  0.037210
15  0.356670
16  0.095604
17  0.210586
18  0.019654
19  0.301963
20  0.097882
21  0.138396
22  0.218453
23  0.197141
24  0.270707

### AUC curve on training set
There is no dependent variable in test set.

### Logistic regresion area under curve

In [64]:
yTrainClean = dfTrainClean["y"].to_numpy()
print(yTrainClean)
falsePositiveRate,truePositiveRate,threshold = metrics.roc_curve(yTrainClean,yProbTrainLog[:,1])

rocAucScore = metrics.roc_auc_score(yTrainClean,yProbTrainLog[:,1])
print("auc score: %s" % rocAucScore)
logLoss = log_loss(yTrainClean,yProbTrainLog[:,1])
print("log loss: %s" % logLoss)

[0. 1. 1. ... 0. 0. 0.]
auc score: 0.7527958790394674
log loss: 0.35991762698348395


### Neural networks area under curve

In [65]:
yTrainClean = dfTrainClean["y"].to_numpy()
print(yTrainClean)
falsePositiveRate,truePositiveRate,threshold = metrics.roc_curve(yTrainClean,yProbTrainNN[:,1])

rocAucScore = metrics.roc_auc_score(yTrainClean,yProbTrainNN[:,1])
print("auc score: %s" % rocAucScore)
logLoss = log_loss(yTrainClean,yProbTrainNN[:,1])
print("log loss: %s" % logLoss)

[0. 1. 1. ... 0. 0. 0.]
auc score: 0.8298681271646436
log loss: 0.31671234359547035
